### PR TITLE
Optimize download process

### LINKED ISSUE #876: Optimize upload and download of large files
https://blob-chart.zus.network/?path=instance1-test3&id=450000x8-EC51

### PR DESCRIPTION
**Pull request by din-mukhammed**
### Changes
Changed the download process to handle downloading and readmarker submission separately, parallelized downloading of blocks.

[Related blobber PR](https://github.com/0chain/blobber/pull/1047)

**LATEST RESULTS:**
_(Downloading 8 files 450MB each)_
Screenshot from 2023-05-03 23-59-43
Screenshot from 2023-05-03 23-59-33
Screenshot from 2023-05-03 23-59-27
Screenshot from 2023-05-03 23-59-18
Screenshot from 2023-05-03 23-59-09
Screenshot from 2023-05-03 23-58-47

### Fixes
-

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/gosdk/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- blobber: https://github.com/0chain/blobber/pull/1047
- 0chain:
- system_test: https://github.com/0chain/system_test/pull/671
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: zboxcore/sdk/downloadworker.go [hunk 1]
```diff
@@ -422,18 +421,31 @@ func (req *DownloadRequest) processDownload(ctx context.Context) {
...
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		req.errorCB(err, remotePathCB)
+		return
+	}
+
+	for _, data := range res {
 		if req.isDownloadCanceled {
```
Comment (lpoli) [inline zboxcore/sdk/downloadworker.go]: This condition should be checked in above block

**Review by lpoli — DISMISSED** [review]
Download is not optimized at all. We should also use readcounter quota system as @dabasov suggested. Also the blockdownloadworker.go should be changed as it makes download request sequential with the use of channel.

For upload I think @Roshan-Mehta 's work will optimize it but please also check for bath upload feature to increase upload size and use it in loadtest.

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | 878-optimize-download  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Review by dabasov — APPROVED** [review]
lgtm

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | 878-optimize-download  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | pr-1047-f1029ad3  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | pr-1047  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | optimize-download-fix  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | 878-optimize-download  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | optimize-download-fix  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | 878-optimize-download  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | optimize-download-fix  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | optimize-download-fix  |
| gosdk  | optimize-download-upload  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | 878-optimize-download  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Pull request merged**